### ISSUE TITLE
An object may not be initialized when vcc_err_unref is off

### ISSUE THREAD

**Issue by Dridi** [varnishcache/varnish-cache#2088]
It is possible to reach the end of `vcl_init` without initializing all objects when `vcc_err_unref` is off. It happens when a `new` statement is in a subroutine that is never called.
## Expected Behavior

An error.
## Current Behavior

A panic.
## Possible Solution

When a `new` statement is evaluated, check that the object is `NULL` if that is not already the case.

When `vcl_init` is done, check that all objects are non NULL.
## Steps to Reproduce (for bugs)

```
varnishtest "unref panic"

varnish v1 -arg "-p vcc_err_unref=off" -vcl {
    import directors;

    backend dummy { .host = "${bad_ip}"; .port = "9090"; }

    sub unrefd {
        new rr = directors.round_robin();
    }

    sub vcl_recv {
        set req.backend_hint = rr.backend();
        return (synth(200));
    }
} -start

client c1 {
    txreq
    rxresp
    expect resp.status == 200
} -run
```

**Comment by gquintard** [comment]
It's actually worse than that, ``vcc_err_unref`` doesn't need to be set:

```
varnishtest "Uninitialized object"

server s1 {}

varnish v1 -vcl+backend {
	import directors;

	sub vcl_init {
		if (0 ==1) {
			new dir = directors.round_robin();
		}
	}

	sub vcl_recv {
		set req.backend_hint = dir.backend();
	}
} -start

client c1 {
	txreq
	rxresp
} -run
```

**Comment by Dridi** [comment]
What I don't say in both "possible solutions" is that I'd gracefully fail the `vcl_init` in both cases.

**Comment by Dridi** [comment]
I found another interesting thing today: if your vmod object fails to initialize and wants to signal it with `VRT_handling(ctx, VCL_RET_FAIL)`, you still can't let it `NULL` as it will make the rest of the VCL unsafe for method calls. And possibly your destructor if it uses `CHECK_OBJ_NOTNULL` to check the pointer.

As part of a potential fix I have in mind, a `new` statement returning with a `NULL` object should immediately fail and leave `vcl_init`.

**Comment by Dridi** [comment]
After a long discussion on IRC today:

>13:56 < phk> ok, first level decision:  if VMODs return NULL on object creation, vcl_init{} fails, and yes, they should be able to explain why.
[...]
14:06 < phk> and one at the end of vcl_init to see that all objects did get a non-NULL value
[...]
14:15 < phk> summary:  new foo = bar() fails vcl_init() if bar() returns NULL.

Assigning to myself, I will come up with a patch after #2145 is closed to avoid potential merge conflicts.

**Comment by rezan** [comment]
So I read the conversation on IRC. My question is, why does a NULL object concern us? As long as we call the constructor, if the VMOD wants its object state to be NULL, thats ok. Its just going to get that NULL value on .method calls and in _fini(). So its really upto the VMOD, if it returns NULL on _init(), then it just needs to support that throughout its API. If it wants that to be an error, then it should raise an error explicitly.

The VCC should only raise an error if the constructor is skipped in my opinion.

**Comment by Dridi** [comment]
> The VCC should only raise an error if the constructor is skipped in my opinion.

I can start with detecting missing constructions and we can revisit the `NULL` question later. I'm not keen on the idea of `NULL` being valid state for an object. As I said, after #2145.

**Comment by rezan** [comment]
NULL is valid for PRIV_* pointers, so its best to let the VMOD decide if it wants to support a null state or not (as VMODs currently do for PRIV_*). Otherwise breaking this convention and injecting NULL checks with implicit escape code for objects will greatly complicate things later (see VIP1 and VIP9).

**Comment by bsdphk** [comment]
I've looked at this, and we need to keep (new foo = bla) objects separate from vmod_priv's.

Objects:

I looked at NULL checking objects in VCC and it gets very hairy, because we need to do the check before we embark on an expression where a vmod is called seven levels deep.

I'm also not keen on saying "vmods must deal with this", because that is just a lot of boilerplate code.

There is a middle solution:  Have vmodtool.py produce a C-wrapper function which does the NULL check on the object, and if NULL does VCL_FAIL and returns a predetermined "safe-ish" value depending on the methods return type.

Privs:

Privs should have both payload & free-func pointers NULL until somebody sets them.  If that is not the case, that's a bug we should fix.  If the free-func pointer is non-null, we call it no matter what the payload pointer is.  If free-func pointer is null and payload is not, we call free(3) on payload.

Does that solve the problems ?

**Comment by nigoroll** [comment]
> Does that solve the problems ?

I know someone who'd respond "send patches" to such a question :)

Nevertheless I think that the wrapper should work.

Nils

**Comment by nigoroll** [comment]
continued in #2839

**Comment by bsdphk** [comment]
This was closed with #2839
